util/feed.go

```diff
@@ -0,0 +1,396 @@
+// created, events can be read directly from the Events channel provided by this
+// structure.
+//
+// An example of a typical usage of a Subscription:
+//
+//		subscriber := feed.Subscribe()
+//		for {
+//			select {
+//			case event, ok := <-subscriber.Events:
+//				// Process event...
+//			}
+//		}
+//
+// A Subscription cannot block other Subscriptions to the same feed, and each
+// Subscription will receive all events published by the Feed. A user of a
+// Subscription should not modify Events received over the channel.
+//
+// A Subscription can be closed via the Unsubscribe() method.
+//
+// If the Feed is stopped, the Events channel will be closed after reading all
+// remaining events that were published to the Feed before it was stopped.
+type Subscription struct {
+	Events <-chan interface{}
+	closer chan<- struct{}
+}
+
+// Unsubscribe immediately stops the given Subscriber. After this method is
+// called, the Events channel will be closed.
+func (s *Subscription) Unsubscribe() {
+	close(s.closer)
+}
+
+// A Feed is used to publish a stream of events to a set of Subscribers.  Events
+// are values of arbitrary type, and are received and published as an empty
+// interface.  Each Subscriber will receive, in order, each entry published to
+// the Feed.
+//
+// Entries are published by via the Publish method of the Feed; this cannot be
+// blocked by Subscriber activities and will always return quickly.  Subscribers
+// receive events by reading from their Events channel.
+//
+// The Feed will continue running until its associated Stopper is closed.
+//
+//		func FeedExample() {
+//			stopper := NewStopper()
+//			feed := StartFeed(stopper)
+//			for i := 0; i < 5; i++ {
+//				go StartSubscriber(i, feed)
+//			}
+//			feed.Publish("Event 1"
+//			feed.Publish("Event 2")
+//			stopper.Stop()
+//		}
+//
+//		func StartSubscriber(i int, f *Feed) {
+//			sub := f.Subscribe()
+//			for event := range sub.Events {
+//				fmt.Println("Subscriber %d received event %s", i,
+//					event.(string))
+//			}
+//		}
+//
+// The Feed's non-blocking properties are achieved with a linked channel instead
+// of buffered channels, and thus there is no limit (other than available
+// memory) to how many unread events the Feed can maintain. This is desirable
+// for Feeds which may publish data in quick bursts.
+//
+// The Feed does not keep historical events; individual Subscribers will only
+// receive events published after they subscribe.
+type Feed struct {
+	events  chan interface{}
+	stopper *Stopper
+
+	// subscriberCount counts the total number of subscribers that have
+	// ever been started. activeCount counts the number of subscribers still
+	// active (i.e. that have not been unsubscribed).
+	subscriberCount int
+	activeCount     int
+
+	// these channels are used for synchronizing different subscription
+	// lifecycle tasks.
+	subscribe   chan *publisher
+	unsubscribe chan *publisher
+	closer      chan struct{}
```

Comment:
I should just change "closer" to "closed", since it is closed by the Feed to indicate closure.